Get working

lucksus opened the pull request:
Post create/get roundtrip

closes #14

lucksus commented:
Oh, and I just noticed that this PR has a green CI now which is not true but an effect of taking faucet out.

**DON'T MERGE THIS**

It actually should be failing until we've merged in @zippy's PRs in the HDK and Holochain.

Connoropolous commented:
@lucksus do you want me to just revert the merge of #22 from master? Because that's the real problem, if you actually want to keep the `build_and_test` file... 

If that's the case, I don't quite get it, because you approved that PR.

Connoropolous commented on build_and_test.sh:
```diff
@@ -0,0 +1,18 @@
+#! /bin/bash
+mkdir dist
```
Would you try switching this file to just the following, since we discussed it and it works?
```
#! /bin/bash
hcdev test  | test/node_modules/faucet/bin/cmd.js

Connoropolous replied:
since this is "app-spec" it sets a better example for others to look at I think

lucksus replied:
The `mkdir dist` is necessary when it's running on travis since empty directories are not committed to git. This was the last fix for the CI that I figured out after many hours.

Also I'd prefer to have all those stages separately in the CI and have those separators in the output since we will work with the app-spec-output a lot going forward.

We expect most of the PRs to come to fail the CI until we build the features used here. So people will be looking at the CI output a lot. The use case for `hcdev test` is running tests manually which is different. If we need to have a shell script for the CI anyway I don't see why we need to use hcdev there too.

Connoropolous commented on test/test.js:
```diff
@@ -1,6 +1,6 @@
 const test = require('tape');
 
-const app = Container.loadAndInstantiate("dist/bundle.json")
+const app = Container.loadAndInstantiate("dist/app-spec-rust.hcpkg")
```
If you switch `build_and_test.sh` like I suggest, this needs to be `dist/bundle.json`, not `dist/app-spec-rust.hcpkg`

Connoropolous commented:
hcdev test performs the exact same steps, inlcuding the logging, minus the
big separators.  Including the creation of dist directory if it doesn't
exist.

On Tue, Oct 9, 2018, 10:37 AM Nicolas Luck <notifications@github.com> wrote:

> *@lucksus* commented on this pull request.
> ------------------------------
>
> In build_and_test.sh
> <https://github.com/holochain/app-spec-rust/pull/20#discussion_r223729459>
> :
>
> > @@ -0,0 +1,18 @@
> +#! /bin/bash
> +mkdir dist
>
> The mkdir dist is necessary when it's running on travis since empty
> directories are not committed to git. This was the last fix for the CI that
> I figured out after many hours.
>
> Also I'd prefer to have all those stages separately in the CI and have
> those separators in the output since we will work with the app-spec-output
> a lot going forward.
>
> We expect most of the PRs to come to fail the CI until we build the
> features used here. So people will be looking at the CI output a lot. The
> use case for hcdev test is running tests manually which is different. If
> we need to have a shell script for the CI anyway I don't see why we need to
> use hcdev there too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/holochain/app-spec-rust/pull/20#discussion_r223729459>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABWAYTd78_Hq7iQFfDnPxoKIKqBoXvXWks5ujLSogaJpZM4XHMd->
> .
>

Connoropolous commented:
But ok I see your point about the CI output. I'll cancel my change request.
It was just to raise the discussion point

On Tue, Oct 9, 2018, 10:51 AM Connor Turland <connor.turland@holo.host>
wrote:

> hcdev test performs the exact same steps, inlcuding the logging, minus the
> big separators.  Including the creation of dist directory if it doesn't
> exist.
>
> On Tue, Oct 9, 2018, 10:37 AM Nicolas Luck <notifications@github.com>
> wrote:
>
>> *@lucksus* commented on this pull request.
>> ------------------------------
>>
>> In build_and_test.sh
>> <https://github.com/holochain/app-spec-rust/pull/20#discussion_r223729459>
>> :
>>
>> > @@ -0,0 +1,18 @@
>> +#! /bin/bash
>> +mkdir dist
>>
>> The mkdir dist is necessary when it's running on travis since empty
>> directories are not committed to git. This was the last fix for the CI that
>> I figured out after many hours.
>>
>> Also I'd prefer to have all those stages separately in the CI and have
>> those separators in the output since we will work with the app-spec-output
>> a lot going forward.
>>
>> We expect most of the PRs to come to fail the CI until we build the
>> features used here. So people will be looking at the CI output a lot. The
>> use case for hcdev test is running tests manually which is different. If
>> we need to have a shell script for the CI anyway I don't see why we need to
>> use hcdev there too.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/holochain/app-spec-rust/pull/20#discussion_r223729459>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABWAYTd78_Hq7iQFfDnPxoKIKqBoXvXWks5ujLSogaJpZM4XHMd->
>> .
>>
>

lucksus commented:
Totally. Thank you, @Connoropolous 👍

lucksus commented:
I was trying now but I don't seem understand how to switch between different types in a linear combinator chain.. can you show me, @AshantiMutinta? :)

Connoropolous commented on build_and_test.sh:
```diff
@@ -0,0 +1,18 @@
+echo "DONE."
+echo "RUNNING tests with holoconsole..."
+echo "------------------------------------------------------------------------------------"
+holoconsole test/dist/bundle.js | test/node_modules/faucet/bin/cmd.js
```
@lucksus change this one to `hcshell` too

lucksus commented:
I don't see how one could exchange these nested matches with combinators, @AshantiMutinta.
I've converted it to early returns but @thedavidmeister and I agree that the nested matches are both better readable and less error prone.

I've added a bunch comments though to explain what `get_entry()` returns and how to deal with it.

Connoropolous commented:
I like that lesser nesting too. To me, I've no concern about which one gets written though

lucksus commented:
Yes, I had exactly that in a commit I didn't push. Early returns are hard to debug, especially in sandboxed WASM code, took about an 1h of my life two weeks ago. This list of assignments holds the notion of constructing something, while what really happens here is dealing with different cases of get_entry's return value. The match represents semantically much better what is happening conceptually than this. I tried this because I hoped it would lead me to finding the right sequence of combinators to use here but the point is that this is not about a sequence but branches of a tree. So, that is both the reason why combinators don't apply and why an iterative list of assignments and early returns are not adequate, I think. It's just that in some other languages there are no better alternatives and that is why some of us (me included) are used to that style.

ddd-mtl commented:
> Early returns are hard to debug, especially in sandboxed WASM code, took about an 1h of my life two weeks ago.

Your anecdote doesn't say (nor prove) what is error prone about it.

I get the part about `match` representing better whats happening. I'd suggest doing error cases before normal cases like this:

```Rust
get_post: |post_hash: String| {
    match hdk::get_entry(post_hash) {
        Err(hdk_error) => hdk_error.to_json(),        
        Ok(maybe_entry) => match maybe_entry {
            None => json!({}),                
            Some(entry) => {
                // normal case flow
                match serde_json::from_str(&entry) {
                    Ok(post) => post,
                    Err(err) => json!({"error deserializing post": err.to_string()}),
                }
            }
        }
    }
} 
```

Also, if this is a common pattern, might do a macro for that? Like `unwrap_zome_api_call!`
I'm not comfortable telling our users the best practice is to start your zome code within 5 levels of indentation.

Connoropolous commented:
lol. I was thinking about that too Damien...

I like the error cases first though
